File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/lite/mobiletestkit/legacy/DocumentRequestHandler.java
package com.couchbase.lite.mobiletestkit.legacy;


import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

import com.couchbase.lite.mobiletestkit.Args;
import com.couchbase.lite.Array;
import com.couchbase.lite.Blob;
import com.couchbase.lite.Dictionary;
import com.couchbase.lite.Document;
import com.couchbase.lite.DocumentChange;
import com.couchbase.lite.DocumentChangeListener;
import com.couchbase.lite.MutableDocument;


public class DocumentRequestHandler {
    /* ------------ */
    /* - Document - */
    /* ------------ */
    public MutableDocument create(Args args) {
        String id = args.getString("id");
        Map<String, Object> dictionary = args.getMap("dictionary");

        if (id != null) {
            if (dictionary == null) {
                return new MutableDocument(id);
            }
            else {
                return new MutableDocument(id, dictionary);
            }
        }
        else {
            if (dictionary == null) {
                return new MutableDocument();
            }
            else {
                return new MutableDocument(dictionary);
            }
        }
    }

    public int count(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        return document.count();
    }

    public String getId(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        return document.getId();
    }

    public String getString(Args args) {
        Document document = args.get("document", Document.class);
        String key = args.getString("key");
        return document.getString(key);
    }

    public MutableDocument setString(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        String value = args.getString("value");
        return document.setString(key, value);
    }

    public Number getNumber(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getNumber(key);
    }

    public MutableDocument setNumber(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Number value = args.getNumber("value");
        return document.setNumber(key, value);
    }


    public Integer getInt(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getInt(key);
    }

    public MutableDocument setInt(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Integer value = args.getInt("value");
        return document.setInt(key, value);
    }


    public Long getLong(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getLong(key);
    }

    public MutableDocument setLong(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Long value = args.getLong("value");
        return document.setLong(key, value);
    }


    public Float getFloat(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getFloat(key);
    }

    public MutableDocument setFloat(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Double valDouble = args.getDouble("value");
        float value = valDouble.floatValue();
        return document.setFloat(key, value);
    }


    public Double getDouble(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getDouble(key);
    }

    public MutableDocument setDouble(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        String value = args.getString("value");
        double set_value = Double.parseDouble(value);
        return document.setDouble(key, set_value);
    }

    public Boolean getBoolean(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getBoolean(key);
    }

    public MutableDocument setBoolean(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Boolean value = args.getBoolean("value");
        return document.setBoolean(key, value);
    }


    public Blob getBlob(Args args) {
        Document document = args.get("document", Document.class);
        String key = args.getString("key");
        return document.getBlob(key);
    }

    public MutableDocument setArray(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Array value = args.get("value", Array.class);
        return document.setArray(key, value);
    }

    public MutableDocument toMutable(Args args) {
        Document document = args.get("document", Document.class);
        return document.toMutable();
    }

    public MutableDocument setData(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        Map<String, Object> data = args.getMap("data");
        return document.setData(data);

    }

    public MutableDocument setBlob(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Blob value = args.get("value", Blob.class);
        return document.setBlob(key, value);
    }


    public Date getDate(Args args) {
        Document document = args.get("document", Document.class);
        String key = args.getString("key");
        return document.getDate(key);
    }

    public MutableDocument setDate(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Date value = args.get("value", Date.class);
        return document.setDate(key, value);
    }


    public Array getArray(Args args) {
        Document document = args.get("document", Document.class);
        String key = args.getString("key");
        return document.getArray(key);
    }

    public Dictionary getDictionary(Args args) {
        Document document = args.get("document", Document.class);
        String key = args.getString("key");
        return document.getDictionary(key);
    }

    public MutableDocument setDictionary(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Dictionary value = args.get("value", Dictionary.class);
        return document.setDictionary(key, value);
    }

    public List<String> getKeys(Args args) {
        Document document = args.get("document", Document.class);
        return document.getKeys();
    }

    public Map<String, Object> toMap(Args args) {
        Document document = args.get("document", Document.class);
        return document.toMap();
    }

    public MutableDocument removeKey(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.remove(key);
    }

    public boolean contains(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.contains(key);
    }

    public String documentChange_getDocumentId(Args args) {
        DocumentChange change = args.get("change", DocumentChange.class);
        return change.getDocumentID();
    }

    public String documentChange_toString(Args args) {
        DocumentChange change = args.get("change", DocumentChange.class);
        return change.toString();
    }

    public Object getValue(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        return document.getValue(key);
    }

    public void setValue(Args args) {
        MutableDocument document = args.get("document", MutableDocument.class);
        String key = args.getString("key");
        Object value = args.getString("value");
        document.setValue(key, value);
    }
}

class MyDocumentChangeListener implements DocumentChangeListener {
    private final List<DocumentChange> changes = new ArrayList<>();

    public List<DocumentChange> getChanges() { return changes; }

    @Override
    public void changed(DocumentChange change) { changes.add(change); }
}

File: CBLClient/Apps/JavaRequestHandlersKotlin/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/DocumentRequestHandler.kt
package com.couchbase.mobiletestkit.javacommon.RequestHandler

import com.couchbase.lite.*
import com.couchbase.lite.Array
import com.couchbase.lite.Dictionary
import com.couchbase.mobiletestkit.javacommon.Args
import java.util.*

class DocumentRequestHandler {
    /* ------------ */ /* - Document - */ /* ------------ */
    fun create(args: Args): MutableDocument {
        val id = args.get<String>("id")
        val dictionary = args.get<Map<String, Any>>("dictionary")
        return if (id != null) {
            dictionary?.let { MutableDocument(id, it) } ?: MutableDocument(id)
        } else {
            dictionary?.let { MutableDocument(it) } ?: MutableDocument()
        }
    }

    fun count(args: Args): Int {
        val document = args.get<MutableDocument>("document")
        return document.count()
    }

    fun getId(args: Args): String {
        val document = args.get<Document>("document")
        return document.id
    }

    fun getString(args: Args): String? {
        val document = args.get<Document>("document")
        val key = args.get<String>("key")
        return document.getString(key)
    }

    fun setString(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<String>("value")
        return document.setString(key, value)
    }

    fun getNumber(args: Args): Number {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getNumber(key)!!
    }

    fun setNumber(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Number>("value")
        return document.setNumber(key, value)
    }

    fun getInt(args: Args): Int {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getInt(key)
    }

    fun setInt(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Int>("value")
        return document.setInt(key, value)
    }

    fun getLong(args: Args): Long {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getLong(key)
    }

    fun setLong(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Long>("value")
        return document.setLong(key, value)
    }

    fun getFloat(args: Args): Float {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getFloat(key)
    }

    fun setFloat(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val valDouble = args.get<Double>("value")
        val value = valDouble.toFloat()
        return document.setFloat(key, value)
    }

    fun getDouble(args: Args): Double {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getDouble(key)
    }

    fun setDouble(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<String>("value")
        val set_value = value.toDouble()
        return document.setDouble(key, set_value)
    }

    fun getBoolean(args: Args): Boolean {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getBoolean(key)
    }

    fun setBoolean(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Boolean>("value")
        return document.setBoolean(key, value)
    }

    fun getBlob(args: Args): Blob? {
        val document = args.get<Document>("document")
        val key = args.get<String>("key")
        return document.getBlob(key)
    }

    fun setArray(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Array>("value")
        return document.setArray(key, value)
    }

    fun toMutable(args: Args): MutableDocument {
        val document = args.get<Document>("document")
        return document.toMutable()
    }

    fun setData(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val data = args.get<Map<String, Any>>("data")
        return document.setData(data)
    }

    fun setBlob(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Blob>("value")
        return document.setBlob(key, value)
    }

    fun getDate(args: Args): Date? {
        val document = args.get<Document>("document")
        val key = args.get<String>("key")
        return document.getDate(key)
    }

    fun setDate(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Date>("value")
        return document.setDate(key, value)
    }

    fun getArray(args: Args): Array? {
        val document = args.get<Document>("document")
        val key = args.get<String>("key")
        return document.getArray(key)
    }

    fun getDictionary(args: Args): Dictionary? {
        val document = args.get<Document>("document")
        val key = args.get<String>("key")
        return document.getDictionary(key)
    }

    fun setDictionary(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Dictionary>("value")
        return document.setDictionary(key, value)
    }

    fun getKeys(args: Args): List<String> {
        val document = args.get<Document>("document")
        return document.keys
    }

    fun toMap(args: Args): Map<String, Any> {
        val document = args.get<Document>("document")
        return document.toMap()
    }

    fun removeKey(args: Args): MutableDocument {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.remove(key)
    }

    operator fun contains(args: Args): Boolean {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.contains(key)
    }

    fun documentChange_getDocumentId(args: Args): String {
        val change = args.get<DocumentChange>("change")
        return change.documentID
    }

    fun documentChange_toString(args: Args): String {
        val change = args.get<DocumentChange>("change")
        return change.toString()
    }

    fun getValue(args: Args): Any {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        return document.getValue(key)!!
    }

    fun setValue(args: Args) {
        val document = args.get<MutableDocument>("document")
        val key = args.get<String>("key")
        val value = args.get<Any>("value")
        document.setValue(key, value)
    }
}

internal class MyDocumentChangeListener : DocumentChangeListener {
    private val changes: MutableList<DocumentChange>? = null
    fun getChanges(): List<DocumentChange>? {
        return changes
    }

    override fun changed(change: DocumentChange) {
        changes!!.add(change)
    }
}